dbms/src/Functions/FunctionsString.cpp

```diff
@@ -5022,52 +5027,410 @@ class FunctionPosition : public IFunction
+
+        for (size_t i = 0; i < row_num; i++)
+        {
+            size_t col0_str_len = col0_offsets[i] - prev_col0_str_offset - 1;
+
+            if unlikely (col0_str_len == 0)
+            {
+                res[i] = 1;
+            }
+            else
+            {
+                const StringRef & col0_collation_str = collator->sortKeyNoTrim(
+                    reinterpret_cast<const char *>(&col0_data[prev_col0_str_offset]),
+                    col0_str_len,
+                    col0_container);
+                void * res_start = memmem(
+                    col1_collation_str.data,
+                    col1_collation_str.size,
+                    col0_collation_str.data,
+                    col0_collation_str.size);
+
+                if (res_start == nullptr)
+                    res[i] = 0;
+                else
+                {
+                    size_t pos = reinterpret_cast<const char *>(res_start) - col1_collation_str.data;
+                    res[i] = 1 + getPositionWithCollationString(pos, lens);
+                }
+            }
+            prev_col0_str_offset = col0_offsets[i];
+        }
+    }
+
+    void constVector(
+        const String & col0_str,
+        const ColumnString::Chars_t & col1_data,
+        const ColumnString::Offsets & col1_offsets,
+        PaddedPODArray<Int64> & res) const
+    {
+        size_t row_num = col1_offsets.size();
+        ColumnString::Offset prev_col1_str_offset = 0;
+
+        String col0_container;
+        String col1_container;
+        std::vector<size_t> lens;
+        StringRef col0_collation_str;
+
+        col0_collation_str = collator->sortKeyNoTrim(col0_str.c_str(), col0_str.size(), col0_container);
+
+        for (size_t i = 0; i < row_num; i++)
+        {
+            size_t col1_str_len = col1_offsets[i] - prev_col1_str_offset - 1;
+
+            const StringRef & col1_collation_str = collator->convert(
+                reinterpret_cast<const char *>(&col1_data[prev_col1_str_offset]),
+                col1_str_len,
+                col1_container,
+                &lens);
+
+            void * res_start = memmem(
+                col1_collation_str.data,
+                col1_collation_str.size,
+                col0_collation_str.data,
+                col0_collation_str.size);
+
+            if (res_start == nullptr)
+                res[i] = 0;
+            else
+            {
+                size_t pos = reinterpret_cast<const char *>(res_start) - col1_collation_str.data;
+                res[i] = 1 + getPositionWithCollationString(pos, lens);
+            }
+
+            prev_col1_str_offset = col1_offsets[i];
+        }
+    }
+
+private:
+    static Int64 getPositionWithCollationString(size_t pos, const std::vector<size_t> & lens)
+    {
+        Int64 actual_pos = 0;
+        size_t idx = 0;
+        while (pos > 0)
+        {
+            actual_pos++;
+            pos -= lens[idx];
```

Comment:
pos is size_t, subtract and compare with 0 is dangerous, maybe we can use signed integer to avoid this.

Reply:
> pos is size_t, subtract and compare with 0 is dangerous, maybe we can use signed integer to avoid this.

done